### PR TITLE
Support signature polymorphic methods (`MethodHandle` and `VarHandle`)

### LINKED ISSUE #11332: Support calling Java 9+ @PolymorphicSignature methods (needed to use MethodHandle and VarHandle)
See https://github.com/scala/scala/pull/4139 and https://github.com/scala/scala/pull/5594

### PR DESCRIPTION
**Pull request by dwijnand**
fixes #11332

### CODE REVIEW

**Comment by SethTisue** [comment]
@sjrd we were hoping you could review the TASTy changes

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 1]
```diff
@@ -485,6 +485,9 @@ class Definitions {
   }
   def NullType: TypeRef = NullClass.typeRef
 
+  @tu lazy val VoidType: TypeRef = requiredClassRef("java.lang.Void")
+  def VoidClass(using Context) = VoidType.symbol.asClass
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:489]: These changes don't seem to be relevant to the rest of the PR.

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 2]
```diff
@@ -413,6 +413,14 @@ class TreePickler(pickler: TastyPickler) {
                 writeByte(if name.isTypeName then SELECTtpt else SELECT)
                 pickleNameAndSig(name, sig, ename)
                 pickleTree(qual)
+              else if defn.wasPolymorphicSignature(tree.symbol) then
+                writeByte(SELECTinPoly)
+                withLength {
+                  pickleNameAndSig(name, tree.symbol.signature, ename)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:419]: Is `tree.symbol.signature` meaningful here? Isn't `tree.symbol` one of those automatically-created `Symbol`s, and hence `signature` would be the polymorphized one, from which we cannot find the symbol again?

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 3]
```diff
@@ -1142,6 +1141,43 @@ class TreeUnpickler(reader: TastyReader,
...
+        val denot =
+          val d = ownerTpe.decl(name).atSignature(sig, target)
+          (if !d.exists then lookupInSuper else d).asSeenFrom(prefix)
+
+        makeSelect(qual, name, denot)
+
+      def readSelectInPoly(): Select =
+        val tree = readSelectIn()
+        val info = readType()
+        tree.withType(tree.symbol.copy(info = info).termRef)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1179]: I find it suspicious that we're creating "member" symbols on the fly like that, without adding them to their owner. I guess this is what nsc does?

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: I'm not 100% sure because I can't tell where cloned symbols are added to their owners in nsc.  Our original idea was that we didn't want to run into overload resolution issues by entering it.

**Comment by smarter** [comment]
> @dwijnand removed the needs-minor-release label

If you have to tweak the unpickler to get the newly pickled stuff to work, then it's still a breaking tasty change and will still require a minor release. Also I don't understand what the changes you're proposing to the unpickler do, so it's not clear to me if it's a better idea than adding a new tag?

**Comment by dwijnand** [comment]
It's peaking through a Typed we added to get the result type, which together with the types of the arguments we can use to do the same symbol copying we do during typing, where we used the args and the prototype.

At the meeting on Monday we discussed whether it was better to special case with a tag or special-case with existing constructs, and decided it might be good to do it without a new tag.

**Comment by sjrd** [comment]
Whether or not it's a new tag, it *is* new semantics in tasty that an existing reader (including earlier versions of the compiler) won't be able to read. So it must be a new minor version of tasty and of the compiler.

Given that, you have a choice between:
* a tag and straightforward code, or
* no tag and complicated logic to read that actually interprets a *combo* of tags as semantically another tag.

I think it's pretty clear that the new tag is preferable.

---

Moreover, I've been thinking about how this should be handled in tasty-query. I really don't want to create fake symbols, since that doesn't match a user's intuition about the language. Because of that, I think in tasty it would make more sense to have a dedicated `APPLY`-like node. One that carries a regular `TermRef` and `args`, *plus* a specific `MethodType`. That wouldn't prevent from reading that into a magical symbol in the compiler, of course.

File: compiler/src/dotty/tools/dotc/transform/Recheck.scala [hunk 4]
```diff
@@ -212,7 +212,9 @@ abstract class Recheck extends Phase, SymTransformer:
       mt.instantiate(argTypes)
 
     def recheckApply(tree: Apply, pt: Type)(using Context): Type =
-      recheck(tree.fun).widen match
+      // reuse the tree's type if the function was a signature polymorphic method, instead of rechecking it
+      val tpe = if tree.fun.symbol.originalSignaturePolymorphic.exists then tree.fun.tpe else recheck(tree.fun)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/transform/Recheck.scala:216]: Shouldn't we at least `recheck(tree.fun.qual)`? It seems we're losing an entire subtree here.

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 5]
```diff
@@ -426,6 +426,13 @@ class TreePickler(pickler: TastyPickler) {
             writeByte(THROW)
             pickleTree(args.head)
           }
+          else if fun.symbol.originalSignaturePolymorphic.exists then
+            writeByte(APPLYsigpoly)
+            withLength {
+              pickleTree(fun)
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:432]: Shouldn't `originalSignaturePolymorphic` be involved here, somehow?

**Review by sjrd — COMMENTED** [review]
Very nice. IMO it's much better now :)

I have two comments which are more questions, really.

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala (hunk 5 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:432]: Like how?  The function Select will be pickled as a SELECTin with the name "invokeExact" and the "in" as MethodHandle, which during unpickling will pick the original method and that's why we have to fix it up.

**Comment by dwijnand** [comment]
@sjrd Had to rebase this.  Any chance you can give it a last look please?

**Comment by sjrd** [comment]
Yes, sorry. I will do that tomorrow afternoon.

**Review by sjrd — APPROVED** [review]

File: project/MiMaFilters.scala [hunk 6]
```diff
@@ -22,4 +22,7 @@ object MiMaFilters {
     ProblemFilters.exclude[MissingClassProblem]("scala.caps$Pure"),
     ProblemFilters.exclude[MissingClassProblem]("scala.caps$unsafe$"),
   )
+  val TastyCore: Seq[ProblemFilter] = Seq(
+    ProblemFilters.exclude[MissingMethodProblem]("dotty.tools.tasty.TastyFormat.APPLYsigpoly"),
```
Comment (smarter) [inline project/MiMaFilters.scala:26]: I think this isn't necessary on the main branch now that the base version is 3.3.0, mima shouldn't complain about new APIs.

Comment (dwijnand) [reply project/MiMaFilters.scala:26]: Ah true

File: tests/run/i11332.scala [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  def main(args: Array[String]): Unit =
+    val l = MethodHandles.lookup()
+    val self    = new Foo()
+    val mhNeg   = l.findVirtual(classOf[Foo], "neg", methodType(classOf[Int], classOf[Int]))
+    val mhRev   = l.findVirtual(classOf[Foo], "rev", methodType(classOf[String], classOf[String]))
+    val mhOverL = l.findVirtual(classOf[Foo], "over", methodType(classOf[String], classOf[Long]))
+    val mhOverI = l.findVirtual(classOf[Foo], "over", methodType(classOf[String], classOf[Int]))
+    val mhUnit  = l.findVirtual(classOf[Foo], "unit", methodType(classOf[Unit], classOf[String]))
+    val mhObj   = l.findVirtual(classOf[Foo], "obj", methodType(classOf[Any], classOf[String]))
+    val mhCL    = l.findStatic(classOf[ClassLoader], "getPlatformClassLoader", methodType(classOf[ClassLoader]))
```
Comment (smarter) [inline tests/run/i11332.scala:24]: This broke the CI because some of our jobs are running on Java 8 which does not define this method:
https://github.com/lampepfl/dotty/actions/runs/3524847656/jobs/5910798965

Comment (dwijnand) [reply tests/run/i11332.scala:24]: https://github.com/lampepfl/dotty/pull/16394
